### PR TITLE
Java: Promote Unsafe resource loading in Android WebView from experimental

### PR DESCRIPTION
**Pull request by atorralba**
PR to promote the "Unsafe resource loading in Android WebView" query created in #3706

## Changes

* Existing files were moved out of `experimental`
* The `UnsafeAndroidAccess.qll` file was created. It contains the classes and predicates used by the main query, but they were refactored to use the CSV sink model and distinguish between an XSS sink and a cross origin sink.
* Removed the `WebView.loadUrl` XSS sink from the XSS library, since it's now being handled better in this query.
* Added tests using `InlineExpectationsTest`. Also added some Android stubs.

## Evaluation

TODO

## Considerations

Currently, the query will be throwing two false positives in the test when concatenating user input to an existing (safe) URL. This gets fixed by using the `RequestForgerySanitizer` in https://github.com/github/codeql/pull/5587, but first that PR needs to be merged.

### CODE REVIEW

File: java/ql/test/query-tests/security/CWE-749/app/UnsafeAndroidAccess.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+			@Override
+			public boolean shouldOverrideUrlLoading(WebView view, String url) {
+				view.loadUrl(url);
+				return true;
+			}
+		});
+
+		String thisUrl = getIntent().getStringExtra("url");
+		wv.loadUrl(thisUrl); // $hasUnsafeAndroidAccess
+		wv.loadUrl("https://www.mycorp.com/" + thisUrl); // $ SPURIOUS: hasUnsafeAndroidAccess // Safe, needs sanitizer
```
Comment (intrigus-lgtm) [inline java/ql/test/query-tests/security/CWE-749/app/UnsafeAndroidAccess.java:73]: Same here.

**Comment by intrigus-lgtm** [comment]
https://github.com/github/codeql/blob/dcee1daa312877ed0d419dc22bf963d7a151b32d/java/ql/src/Security/CWE/CWE-749/UnsafeAndroidAccess.qhelp#L26
This line could be removed from the query help, as query help generation will automatically pick it up from here:
https://github.com/github/codeql/blob/be50e8f30cdec349eb5a6f680336561d9f51145f/java/ql/src/Security/CWE/CWE-749/UnsafeAndroidAccess.ql#L10-L11

**Comment by atorralba** [comment]
@github/docs-content-codeql please review the qhelp file. Even though changes aren't introduced in this PR, it wasn't reviewed when this query was merged to experimental.

File: java/ql/src/semmle/code/java/security/UnsafeAndroidAccess.qll [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+private import semmle.code.java.dataflow.ExternalFlow
+
+/**
+ * A sink that represents a method that fetches a web resource in Android.
+ *
+ * Extend this class to add your own Unsafe Resource Fetching sinks.
+ */
+abstract class UrlResourceSink extends DataFlow::Node {
+  /**
+   * Returns a description of this vulnerability,
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/security/UnsafeAndroidAccess.qll:17]:
```suggestion
   * Gets a description of this vulnerability.
```

File: java/ql/src/semmle/code/java/security/UnsafeAndroidAccess.qll [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+   */
+  abstract string getSinkType();
+}
+
+/**
+ * Cross-origin access enabled resource fetch.
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/security/UnsafeAndroidAccess.qll:23]:
```suggestion
 * A cross-origin access enabled resource fetch.
```

File: java/ql/src/semmle/code/java/security/UnsafeAndroidAccess.qll [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  JavaScriptEnabledUrlResourceSink() {
+    exists(MethodAccess loadUrl, VarAccess webviewVa, MethodAccess getSettingsMa, Variable v |
+      loadUrl.getArgument(0) = this.asExpr() and
+      loadUrl.getMethod() instanceof WebViewLoadUrlMethod and
+      loadUrl.getQualifier() = webviewVa and
+      getSettingsMa.getMethod() instanceof WebViewGetSettingsMethod and
+      webviewVa.getVariable().getAnAccess() = getSettingsMa.getQualifier() and
+      v.getAnAssignedValue() = getSettingsMa and
+      isJSEnabled(v)
+    )
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/security/UnsafeAndroidAccess.qll:57]: Parts of this logic is repeated in the other charpred above in `CrossOriginUrlResourceSink`. Pull out the shared bits in a predicate, e.g.:
```ql
/**
 * Holds if a `WebViewLoadUrlMethod` method is called with the given `urlArg` on a
 * WebView with settings stored in `settings`.
 */
private predicate webViewLoadUrl(Expr urlArg, Variable settings) {
  exists(MethodAccess loadUrl, Variable webview, MethodAccess getSettings |
    loadUrl.getArgument(0) = urlArg and
    loadUrl.getMethod() instanceof WebViewLoadUrlMethod and
    loadUrl.getQualifier() = webview.getAnAccess() and
    getSettings.getMethod() instanceof WebViewGetSettingsMethod and
    webview.getAnAccess() = getSettingsMa.getQualifier() and
    settings.getAnAssignedValue() = getSettings
  )
}
```
Then this charpred can be written as
```suggestion
    exists(Variable settings |
      webViewLoadUrl(this.asExpr(), settings) and
      isJSEnabled(settings)
    )
```
And the above charpred
```ql
  CrossOriginUrlResourceSink() {
    exists(Variable settings, MethodAccess ma, MethodAccess getSettingsMa |
      webViewLoadUrl(this.asExpr(), settings) and
      ma.getMethod() instanceof CrossOriginAccessMethod and
      ma.getArgument(0).(BooleanLiteral).getBooleanValue() = true and
      ma.getQualifier() = settings.getAnAccess()
    )
  }
```

File: java/ql/src/semmle/code/java/security/UnsafeAndroidAccess.qll [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  }
+
+  override string getSinkType() { result = "user input vulnerable to XSS attacks" }
+}
+
+/**
+ * Methods allowing any-local-file and cross-origin access in the WebSettings class
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/security/UnsafeAndroidAccess.qll:64]:
```suggestion
 * A method allowing any-local-file and cross-origin access in the WebSettings class.
```

File: java/ql/src/semmle/code/java/security/UnsafeAndroidAccess.qll [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+ */
+private class CrossOriginAccessMethod extends Method {
+  CrossOriginAccessMethod() {
+    this.getDeclaringType() instanceof TypeWebSettings and
+    this.hasName(["setAllowUniversalAccessFromFileURLs", "setAllowFileAccessFromFileURLs"])
+  }
+}
+
+/**
+ * `setJavaScriptEnabled` method for the webview
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/security/UnsafeAndroidAccess.qll:74]:
```suggestion
 * The `setJavaScriptEnabled` method for the webview.
```

File: java/ql/src/semmle/code/java/security/UnsafeAndroidAccess.qll [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+ */
+private class AllowJavaScriptMethod extends Method {
+  AllowJavaScriptMethod() {
+    this.getDeclaringType() instanceof TypeWebSettings and
+    this.hasName("setJavaScriptEnabled")
+  }
+}
+
+/**
+ * Holds if a call to `v.setJavaScriptEnabled(true)` exists
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/security/UnsafeAndroidAccess.qll:84]:
```suggestion
 * Holds if a call to `v.setJavaScriptEnabled(true)` exists.
```

File: java/ql/src/semmle/code/java/security/UnsafeAndroidAccessQuery.qll [hunk 8]
```diff
@@ -0,0 +1,22 @@
+/** Provides taint tracking configurations to be used in Unsafe Resource Fetching queries. */
+
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.security.RequestForgery
+import semmle.code.java.security.UnsafeAndroidAccess
+
+/**
+ * Taint configuration tracking flow from untrusted inputs to a resource fetching call.
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/security/UnsafeAndroidAccessQuery.qll:10]:
```suggestion
 * A taint configuration tracking flow from untrusted inputs to a resource fetching call.
```

**Review by felicitymay — COMMENTED** [review]
I can't comment on the `.qhelp` file or raise a PR against this PR with suggestions.

Suggested update to the `.qhelp` file:

```
<!DOCTYPE qhelp PUBLIC "-//Semmle//qhelp//EN" "qhelp.dtd">
<qhelp>

  <overview>
    <p>Android WebViews that allow loading URLs controlled by external inputs, and whose JavaScript interface is enabled, are potentially vulnerable to cross-site scripting and sensitive resource disclosure attacks.</p>
    <p>A <code>WebView</code> whose <code>WebSettings</code> object has called <code>setAllowFileAccessFromFileURLs(true)</code> or <code>setAllowUniversalAccessFromFileURLs(true)</code> must not load any untrusted web content.</p>
    <p>Enabling these settings allows malicious scripts loaded in a <code>file://</code> context to launch cross-site scripting attacks, either accessing arbitrary local files including WebView cookies, session tokens, private app data or even credentials used on arbitrary web sites.</p>
    <p>This query detects the following two scenarios:</p>
    <ol>
      <li>Vulnerability introduced by WebViews with JavaScript enabled and remote inputs allowed.</li>
      <li>A more severe vulnerability when "allow cross-origin resource access" is also enabled. This setting was deprecated in API level 30 (Android 11), but most devices are still affected, especially since some Android phones are updated slowly or no longer updated at all.</li>
    </ol>
  </overview>

  <recommendation>
    <p>Only allow trusted web content to be displayed in WebViews when JavaScript is enabled. Disallow cross-origin resource access in WebSettings to reduce the attack surface.</p>
  </recommendation>

  <example>
    <p>The following example shows both 'BAD' and 'GOOD' configurations. In the 'BAD' configuration, JavaScript and the allow access setting are enabled and URLs are loaded from externally controlled inputs. In the 'GOOD' configuration, JavaScript is disabled or only trusted web content is allowed to be loaded.</p>
    <sample src="UnsafeAndroidAccess.java" />
  </example>

  <references>
    <li>
      Google Help: <a href="https://support.google.com/faqs/answer/7668153?hl=en">Fixing a File-based XSS Vulnerability</a>
    </li>
    </li>
      OWASP: <a href="https://github.com/OWASP/owasp-mstg/blob/master/Document/0x05h-Testing-Platform-Interaction.md#testing-javascript-execution-in-webviews-mstg-platform-5">Testing JavaScript Execution in WebViews (MSTG-PLATFORM-5)</a>
    </li>
    </li>
      OWASP: <a href="https://github.com/OWASP/owasp-mstg/blob/master/Document/0x05h-Testing-Platform-Interaction.md#testing-webview-protocol-handlers-mstg-platform-6">Testing WebView Protocol Handlers (MSTG-PLATFORM-6)</a>
    </li>
  </references>
</qhelp>
```

**Comment by atorralba** [comment]
@felicitymay @aschackmull thank you for the reviews! I applied all your suggestions, we only need to make a decision about the CodeQL packages mention.

File: java/ql/src/Security/CWE/CWE-749/UnsafeAndroidAccess.qhelp [hunk 9]
```diff
@@ -0,0 +1,35 @@
+<!DOCTYPE qhelp PUBLIC "-//Semmle//qhelp//EN" "qhelp.dtd">
+<qhelp>
+
+  <overview>
+    <p>Android WebViews that allow loading URLs controlled by external inputs, and whose JavaScript interface is enabled, are potentially vulnerable to cross-site scripting and sensitive resource disclosure attacks.</p>
```
Comment (felicitymay) [inline java/ql/src/Security/CWE/CWE-749/UnsafeAndroidAccess.qhelp:5]:
```suggestion
    <p>Android WebViews that allow externally controlled URLs to be loaded, and whose JavaScript interface is enabled, are potentially vulnerable to cross-site scripting and sensitive resource disclosure attacks.</p>
```

File: java/ql/src/Security/CWE/CWE-749/UnsafeAndroidAccess.qhelp [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    <p>A <code>WebView</code> whose <code>WebSettings</code> object has called <code>setAllowFileAccessFromFileURLs(true)</code> or <code>setAllowUniversalAccessFromFileURLs(true)</code> must not load any untrusted web content.</p>
+    <p>Enabling these settings allows malicious scripts loaded in a <code>file://</code> context to launch cross-site scripting attacks, either accessing arbitrary local files including WebView cookies, session tokens, private app data or even credentials used on arbitrary web sites.</p>
```
Comment (felicitymay) [inline java/ql/src/Security/CWE/CWE-749/UnsafeAndroidAccess.qhelp:7]:
```suggestion
    <p>Enabling these settings allows malicious scripts loaded in a <code>file://</code> context to launch cross-site scripting attacks, accessing arbitrary local files including WebView cookies, session tokens, private app data or even credentials used on arbitrary web sites.</p>
```

**Review by felicitymay — COMMENTED** [review]
@atorralba - thanks for making those text changes. 

Now that I can see my suggestions in place in the artifact, I spotted a few places where the meaning could be clearer and have added a few more comments. When these changes are added, I'm happy for this PR to be merged from the docs point of view.

**Review by aschackmull — APPROVED** [review]

**Pull request merged**